### PR TITLE
Fix fad types

### PR DESCRIPTION
**Pull request by mperego**
This branch 
1. makes Sacado an optional dependency for Intrepid2 (thanks Nathan Ellingwood for helping with this),
2. fixes a few issues when using different types (double and Fad) for the evaluation points and the evaluated basis functions.
3. allocates dynamically the buffer required for high order basis functions (thanks to Kyungjoo for suggesting a way to do this).
4. disables the computation of second and higher derivativs derivatives for high order basis functions on triangles and Tets. It is in fact very hard to handle the nested fad types recursively computed used for high order derivatives, and we do not know of any use case.
4. modifies/adds several tests to Intrepid2 for checking use of high-order bases with fad types. (derivative values are set to zero and not checked). Tests with fad types are currently disabled when UVM is not enabled.  
4. clean the Intrepid2 folder removing old core folder and moving "refactor" folder content in the intrepid2 folder.

### CODE REVIEW

File: packages/panzer/dof-mgr/test/field_pattern/tFieldAggPattern2.cpp [hunk 1]
```diff
@@ -154,7 +157,8 @@ namespace panzer {
   {
     out << note << std::endl;
 
-    const int poly_A = 3, poly_B = 5;
+    const int poly_A = std::min(3, Intrepid2::Parameters::MaxOrder),
+              poly_B = std::min(4, Intrepid2::Parameters::MaxOrder);
```
Comment (ndellingwood) [inline packages/panzer/dof-mgr/test/field_pattern/tFieldAggPattern2.cpp]: @mperego We noticed the '4' on this line should be '5' during our review

Comment (mperego) [reply packages/panzer/dof-mgr/test/field_pattern/tFieldAggPattern2.cpp]: Thanks! I'll fix this.

**Comment by mperego** [comment]
Kyungjoo suggested a way to dynamically allocate the buffer needed for high order bases functions. I'll try to implement this, so please hold off until I fix this.

**Comment by mperego** [comment]
@ndellingwood, @kyungjoo, I have adressd your requests. Please let me know if I can merge this pull request.
@bathmatt @rppawlo, please let me know if you want to check if this breaks any dependent applications. If I do not hear from you I will try to merge it sometime tomorrow.

**Comment by mperego** [comment]
@kyungjoo-kim , please have a look at commit  33160beedd9b830. I'm still hesitant in exposing the Fad dimesion in Intrepid2, but if you prefer this solution I'm fine with that.

**Pull request merged**